always destroy socket provider on disconnect, and fix type error

trevj opened the pull request:
A new version of the TypeScript compiler highlighted this type error:
https://github.com/uProxy/uproxy-networking/blob/dev/src/net/tcp.ts#L61

Fixing it, I realised two things:
1. `close` throws an error when the socket is already closed
2. we have only been destroying the provider when we explicitly call `close`

I believe it's more appropriate to destroy the provider once we receive an `onDisconnect` event. This may fix:
https://github.com/uProxy/uproxy/issues/1303

@willscott could you also take a tiny quick look at this?

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/uproxy/uproxy-networking/254)

<!-- Reviewable:end -->

trevj committed:
always destroy socket provider on disconnect, and fix type error

bemasc commented on src/integration-tests/tcp/freedom-module.ts:
```diff
@@ -59,12 +59,12 @@ parentModule.on('shutdown', () => {
     server.connectionsQueue.setSyncHandler((connection:tcp.Connection) => {
       client.onceConnected.then(() => {
-        server.shutdown();
+    server.shutdown();
```
This spacing seems ... worse.

trevj replied:
Sorry, I had my diff tool ignoring leading whitespace for some strange reason...

bemasc commented on src/net/tcp.ts:
```diff
@@ -46,20 +46,15 @@ export function endpointOfSocketInfo(info:freedom_TcpSocket.SocketInfo)
+//   freedom['core.tcpsocket']().close
+//
+// This is different from:
+//   freedom['core.tcpsocket'].close
```
Do you mean freedomTcpSocket.close()?
EDIT: I mean, are these two cases backwards?

trevj replied:
Oh god, you're right -- and I looked extra hard at that. I shortened this, hopefully it's clearer now (and a lot shorter).

bemasc commented on src/net/tcp.ts:
```diff
@@ -116,6 +111,18 @@ export class Server {
+    // Since there may be other calls to this provider in progress which will
+    // fail to resolve if we destroy the communication channel now, postpone
+    // doing so until the next iteration of the event loop.
+    setTimeout(() => {
```
This seems incredibly fragile.  Is there a reason that this is actually safe?  Could we structure this to wait for a promise so that it's actually deterministic?

trevj replied:
You're right. Probably happenstance that this works right now. I couldn't think of a simple general solution...let me think about it some more.

trevj replied:
OK, it was kinda fun to try to come up with a non-horrible solution...I'm not sure I have, although the integration tests pass now.

This of course is without changing the guarantee on when `onDisconnect` fires, which we might also consider.

trevj committed:
fix indentation

trevj committed:
simplify and shorten comment for destroyFreedomSocket_

trevj commented:
Thanks for looking at this!

trevj committed:
track number of inflight calls to the socket and only destroy its provider when all have finished

trevj committed:
also have client sockets track their number of inflight calls for socket destruction

trevj commented:
PTAL!

bemasc commented:
<sup>**[src/net/tcp.ts, line 67 [r4]](https://reviewable.io:443/reviews/uproxy/uproxy-networking/254#-JoFamcx71nGv2n-fNWp)** ([raw file](https://github.com/uproxy/uproxy-networking/blob/891c934246ca3869621eaae51e619c5c795c196b/src/net/tcp.ts#L67)):</sup>
How about something like

``` typescript
before();
return f().then((val:T) => {
  after()
  return val;
}, (e:Error) => {
  after();
  throw e;
};
```

That would avoid the ungainly `new Promise` construction.

---

<sup>**[src/net/tcp.ts, line 131 [r4]](https://reviewable.io:443/reviews/uproxy/uproxy-networking/254#-JoFp03zvFFBUzHRYSwV)** ([raw file](https://github.com/uproxy/uproxy-networking/blob/891c934246ca3869621eaae51e619c5c795c196b/src/net/tcp.ts#L131)):</sup>
I think this needs a comment to explain that its purpose is to drive the count to -1, which is a special condition indicating termination.

---

<sup>**[src/net/tcp.ts, line 399 [r4]](https://reviewable.io:443/reviews/uproxy/uproxy-networking/254#-JoFpInWhIN1pyIh9QaX)** ([raw file](https://github.com/uproxy/uproxy-networking/blob/891c934246ca3869621eaae51e619c5c795c196b/src/net/tcp.ts#L399)):</sup>
Binding `this` to `undefined` seems pretty scary to me.

---

<sup>**[src/net/tcp.ts, line 428 [r4]](https://reviewable.io:443/reviews/uproxy/uproxy-networking/254#-JoFpO8pcbDLxB-B8Mip)** ([raw file](https://github.com/uproxy/uproxy-networking/blob/891c934246ca3869621eaae51e619c5c795c196b/src/net/tcp.ts#L428)):</sup>
Same here

---

<sup>**[src/net/tcp.ts, line 453 [r4]](https://reviewable.io:443/reviews/uproxy/uproxy-networking/254#-JoFpS2d3-ByOdb-7PBy)** ([raw file](https://github.com/uproxy/uproxy-networking/blob/891c934246ca3869621eaae51e619c5c795c196b/src/net/tcp.ts#L453)):</sup>
Copy and paste makes me sad.  What if the wrapper were its own class, and then `Connection` and `Server` each had one?

---

---

Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy-networking/254)

<!-- Sent from Reviewable.io -->

trevj committed:
simplify wrap async handling and throw before exceptions synchronously

trevj committed:
add comment for inflight termination condition

trevj committed:
sockets should be bound to sockets

trevj commented:
Thanks @bemasc, PTAL!

---

<sup>**[src/integration-tests/tcp/freedom-module.ts, line 62 [r1]](https://reviewable.io:443/reviews/uproxy/uproxy-networking/254#-Jo1EshnVHq9hprN-VxO-r1-62)** ([raw file](https://github.com/uproxy/uproxy-networking/blob/60549e1cebe945a20d5ab6afd9eda1b318aaf8e7/src/integration-tests/tcp/freedom-module.ts#L62)):</sup>
Done.

---

<sup>**[src/net/tcp.ts, line 51 [r1]](https://reviewable.io:443/reviews/uproxy/uproxy-networking/254#-Jo1EshnVHq9hprN-VxP-r1-53)** ([raw file](https://github.com/uproxy/uproxy-networking/blob/60549e1cebe945a20d5ab6afd9eda1b318aaf8e7/src/net/tcp.ts#L51)):</sup>
Done.

---

<sup>**[src/net/tcp.ts, line 67 [r4]](https://reviewable.io:443/reviews/uproxy/uproxy-networking/254#-JoFamcx71nGv2n-fNWp)** ([raw file](https://github.com/uproxy/uproxy-networking/blob/891c934246ca3869621eaae51e619c5c795c196b/src/net/tcp.ts#L67)):</sup>
Right. I did this so any error raised by `after()` would propagate asynchronously but just didn't realised what I had didn't work...changed now. `before()` will just raise eceptions synchronously.

---

<sup>**[src/net/tcp.ts, line 399 [r4]](https://reviewable.io:443/reviews/uproxy/uproxy-networking/254#-JoFpInWhIN1pyIh9QaX)** ([raw file](https://github.com/uproxy/uproxy-networking/blob/891c934246ca3869621eaae51e619c5c795c196b/src/net/tcp.ts#L399)):</sup>
Ah, yes. Probably only works due to how the socket provider is written internally. Fixed, thanks!

---

<sup>**[src/net/tcp.ts, line 453 [r4]](https://reviewable.io:443/reviews/uproxy/uproxy-networking/254#-JoFpS2d3-ByOdb-7PBy)** ([raw file](https://github.com/uproxy/uproxy-networking/blob/891c934246ca3869621eaae51e619c5c795c196b/src/net/tcp.ts#L453)):</sup>
Right. I've extracted this to a separate file, with a test. It's a little tricky because only standalone functions can be parameterised...but I think this is about as clean as I can get it.

---

---

Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy-networking/254)

<!-- Sent from Reviewable.io -->

trevj committed:
extract socket provider call counting to a separate class

bemasc commented:
<sup>**[src/net/counter.ts, line 10 [r6]](https://reviewable.io:443/reviews/uproxy/uproxy-networking/254#-JoVciQClpEb-JwEnIth)** ([raw file](https://github.com/uproxy/uproxy-networking/blob/6669641b0dde507e3afce6853d84e193cbd5fca4/src/net/counter.ts#L10)):</sup>
I think this would be more attractive as a method of `Counter`.  Then `.before` and `.after` could be private.

---

<sup>**[src/net/counter.ts, line 39 [r6]](https://reviewable.io:443/reviews/uproxy/uproxy-networking/254#-JoVdjIOh5idaFFYIUMB)** ([raw file](https://github.com/uproxy/uproxy-networking/blob/6669641b0dde507e3afce6853d84e193cbd5fca4/src/net/counter.ts#L39)):</sup>
I think this pattern needs a comment to explain how it relates to the `this.counter_ < 0` check.  Alternatively, maybe a comment on `counter_` to explain its semantics.

---

<sup>**[src/net/tcp.ts, line 416 [r6]](https://reviewable.io:443/reviews/uproxy/uproxy-networking/254#-JoVdZe0UrO2zoHp3GMw)** ([raw file](https://github.com/uproxy/uproxy-networking/blob/6669641b0dde507e3afce6853d84e193cbd5fca4/src/net/tcp.ts#L416)):</sup>
Do you want to wait for this.counter_.onceDestroyed here?

---

Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy-networking/254)

<!-- Sent from Reviewable.io -->

trevj committed:
onDisconnect handlers should wait until the counter is destroyed to fulfill onceClosed

trevj commented:
<sup>**[src/net/tcp.ts, line 416 [r6]](https://reviewable.io:443/reviews/uproxy/uproxy-networking/254#-JoVdZe0UrO2zoHp3GMw)** ([raw file](https://github.com/uproxy/uproxy-networking/blob/6669641b0dde507e3afce6853d84e193cbd5fca4/src/net/tcp.ts#L416)):</sup>
YES. Excellent point. Done, for server and client sockets. I like it much better now. Thanks.

---

<sup>**[src/net/counter.ts, line 10 [r6]](https://reviewable.io:443/reviews/uproxy/uproxy-networking/254#-JoVciQClpEb-JwEnIth)** ([raw file](https://github.com/uproxy/uproxy-networking/blob/6669641b0dde507e3afce6853d84e193cbd5fca4/src/net/counter.ts#L10)):</sup>
So, I agree -- however, I wanted type safety on the return type and TypeScript only supports this on standalone functions.

I don't think there's any way to move this method to `Counter` without making the return type `any`, which I think would be worse.

---

<sup>**[src/net/counter.ts, line 39 [r6]](https://reviewable.io:443/reviews/uproxy/uproxy-networking/254#-JoVdjIOh5idaFFYIUMB)** ([raw file](https://github.com/uproxy/uproxy-networking/blob/6669641b0dde507e3afce6853d84e193cbd5fca4/src/net/counter.ts#L39)):</sup>
Done.

---

Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy-networking/254)

<!-- Sent from Reviewable.io -->

trevj committed:
explain how discard works

bemasc commented:
<sup>**[src/net/counter.ts, line 10 [r6]](https://reviewable.io:443/reviews/uproxy/uproxy-networking/254#-JoVciQClpEb-JwEnIth)** ([raw file](https://github.com/uproxy/uproxy-networking/blob/6669641b0dde507e3afce6853d84e193cbd5fca4/src/net/counter.ts#L10)):</sup>
Just make `Counter<T>` parametric?

---

Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy-networking/254)

<!-- Sent from Reviewable.io -->

bemasc commented:
<sup>**[src/net/counter.ts, line 10 [r6]](https://reviewable.io:443/reviews/uproxy/uproxy-networking/254#-JoVciQClpEb-JwEnIth)** ([raw file](https://github.com/uproxy/uproxy-networking/blob/6669641b0dde507e3afce6853d84e193cbd5fca4/src/net/counter.ts#L10)):</sup>
Oh, I see; it's used to wrap several different types.  Hmm...

---

Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy-networking/254)

<!-- Sent from Reviewable.io -->

trevj commented:
<sup>**[src/net/counter.ts, line 10 [r6]](https://reviewable.io:443/reviews/uproxy/uproxy-networking/254#-JoVciQClpEb-JwEnIth)** ([raw file](https://github.com/uproxy/uproxy-networking/blob/6669641b0dde507e3afce6853d84e193cbd5fca4/src/net/counter.ts#L10)):</sup>
Right. It works great but only on a standalone function.

---

Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy-networking/254)

<!-- Sent from Reviewable.io -->

bemasc commented:
<sup>**[src/net/counter.ts, line 10 [r6]](https://reviewable.io:443/reviews/uproxy/uproxy-networking/254#-JoVciQClpEb-JwEnIth)** ([raw file](https://github.com/uproxy/uproxy-networking/blob/6669641b0dde507e3afce6853d84e193cbd5fca4/src/net/counter.ts#L10)):</sup>
I just tried this locally.  I think you can do

``` typescript
public wrap = <T>(f:() => Promise<T>) : Promise<T> => {
```

---

Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy-networking/254)

<!-- Sent from Reviewable.io -->

trevj committed:
move wrap function to the counter class

trevj commented:
<sup>**[src/net/counter.ts, line 10 [r6]](https://reviewable.io:443/reviews/uproxy/uproxy-networking/254#-JoVciQClpEb-JwEnIth)** ([raw file](https://github.com/uproxy/uproxy-networking/blob/6669641b0dde507e3afce6853d84e193cbd5fca4/src/net/counter.ts#L10)):</sup>
Oh! You know, I tried something very similar and couldn't get the magic incantation right. Many thanks, fixed.

---

Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy-networking/254)

<!-- Sent from Reviewable.io -->

bemasc commented:
:+1: 

---

Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy-networking/254)

<!-- Sent from Reviewable.io -->

trevj committed:
Merge remote-tracking branch 'origin/dev' into trevj-tcp-close

trevj committed:
bump patch version for tcp socket closing fix

trevj committed:
Merge branch 'dev' into trevj-tcp-close
